module/state_synchronization/execution_data_storer.go

```diff
@@ -109,44 +77,31 @@ func (s *StateDiffStorer) Store(sd *ExecutionStateDiff) (cid.Cid, error) {
 	}
 }
 
-func (s *StateDiffStorer) readBlocks(cids []cid.Cid) (interface{}, error) {
+func (s *ExecutionDataStorer) readBlocks(cids []cid.Cid) (interface{}, error) {
 	buf := &bytes.Buffer{}
-	comp, err := s.compressor.NewReader(buf)
-	if err != nil {
-		return nil, err
-	}
-	dec := s.codec.NewDecoder(comp)
-	var code byte
-	for i, c := range cids {
+	for _, c := range cids {
 		block, err := s.blockWriter.bstore.Get(c)
 		if err != nil {
 			return nil, fmt.Errorf("failed to get block %v from blockstore: %w", c, err)
 		}
-		data := block.RawData()
-		if i == 0 {
-			code = data[0]
-			data = data[1:]
-		}
-		_, _ = buf.Write(data) // never returns error
+		_, _ = buf.Write(block.RawData()) // never returns error
 	}
-	v := getPrototype(code)
-	err = dec.Decode(v)
-	return v, err
+	return s.serializer.deserialize(buf)
 }
 
-// Load loads the ExecutionStateDiff represented by the given CID from the blockstore.
+// Load loads the ExecutionData represented by the given CID from the blockstore.
 // Since blocks are limited to MAX_BLOCK_SIZE bytes, it's possible that the data was
-// stored in multiple chunks, and hence the root CID may point to a block for which
-// the data itself is a list of concatenated CIDs.
-func (s *StateDiffStorer) Load(c cid.Cid) (*ExecutionStateDiff, error) {
+// stored in multiple blocks, and hence the root CID may point to a block for which
+// the data itself represents a recursive list of CIDs.
```

Comment:
So, the implementation is actually intended to handle the general case where:

Starting from the root CID, we load the block which contains a chunk of RawBytes.

These RawBytes may represent an ExecutionData, or it may represent a list of CIDs. In the latter case, we load the blocks from the list and concatenate the data into a new chunk of RawBytes. Then we are back to where we started, because this new RawBytes could either represent an ExecutionData, or another list of CIDs.

So, in theory, this tree of CIDs could grow infinitely deep, and the implementation is supposed to handle this using the for loop. In practice, the tree will be at most 2 levels deep assuming the largest ExecutionData we will ever see is ~400MB.

That's why I used the word "recursive", but if this is confusing I'm open to suggestions on how to reword it